File: 3. Behavioral Design Patterns/Command/Java/CommandExample.java
// Command Pattern Example in Java
// Remote control executes commands on Light and Stereo

interface Command { void execute(); void undo(); }

class Light { void on(){ System.out.println("Light ON"); } void off(){ System.out.println("Light OFF"); } }
class Stereo { void on(){ System.out.println("Stereo ON"); } void off(){ System.out.println("Stereo OFF"); } }

class LightOnCommand implements Command { private final Light l; LightOnCommand(Light l){ this.l = l; } public void execute(){ l.on(); } public void undo(){ l.off(); } }
class LightOffCommand implements Command { private final Light l; LightOffCommand(Light l){ this.l = l; } public void execute(){ l.off(); } public void undo(){ l.on(); } }
class StereoOnCommand implements Command { private final Stereo s; StereoOnCommand(Stereo s){ this.s = s; } public void execute(){ s.on(); } public void undo(){ s.off(); } }
class StereoOffCommand implements Command { private final Stereo s; StereoOffCommand(Stereo s){ this.s = s; } public void execute(){ s.off(); } public void undo(){ s.on(); } }

class RemoteButton { private Command cmd; RemoteButton(Command c){ cmd = c; } void set(Command c){ cmd = c; } void press(){ cmd.execute(); } void pressUndo(){ cmd.undo(); } }

public class CommandExample {
    public static void main(String[] args){
        Light light = new Light(); Stereo stereo = new Stereo();
        RemoteButton btn = new RemoteButton(new LightOnCommand(light));
        btn.press(); btn.pressUndo();
        btn.set(new StereoOnCommand(stereo));
        btn.press(); btn.pressUndo();
    }
}

File: 3. Behavioral Design Patterns/Command/C#/CommandExample.cs
// Command Pattern Example in C#
// Remote control executes commands on Light and Stereo

using System;

namespace CommandExample
{
    interface ICommand { void Execute(); void Undo(); }

    class Light { public void On(){ Console.WriteLine("Light ON"); } public void Off(){ Console.WriteLine("Light OFF"); } }
    class Stereo { public void On(){ Console.WriteLine("Stereo ON"); } public void Off(){ Console.WriteLine("Stereo OFF"); } }

    class LightOnCommand : ICommand { private readonly Light _l; public LightOnCommand(Light l){ _l = l; } public void Execute(){ _l.On(); } public void Undo(){ _l.Off(); } }
    class LightOffCommand : ICommand { private readonly Light _l; public LightOffCommand(Light l){ _l = l; } public void Execute(){ _l.Off(); } public void Undo(){ _l.On(); } }
    class StereoOnCommand : ICommand { private readonly Stereo _s; public StereoOnCommand(Stereo s){ _s = s; } public void Execute(){ _s.On(); } public void Undo(){ _s.Off(); } }
    class StereoOffCommand : ICommand { private readonly Stereo _s; public StereoOffCommand(Stereo s){ _s = s; } public void Execute(){ _s.Off(); } public void Undo(){ _s.On(); } }

    class RemoteButton { private ICommand _cmd; public RemoteButton(ICommand cmd){ _cmd = cmd; } public void Set(ICommand cmd){ _cmd = cmd; } public void Press(){ _cmd.Execute(); } public void PressUndo(){ _cmd.Undo(); } }

    class Program
    {
        static void Main()
        {
            var light = new Light(); var stereo = new Stereo();
            var btn = new RemoteButton(new LightOnCommand(light));
            btn.Press(); btn.PressUndo();
            btn.Set(new StereoOnCommand(stereo));
            btn.Press(); btn.PressUndo();
        }
    }
}
